core/src/main/java/cz/o2/proxima/storage/StreamElement.java

```diff
@@ -158,17 +262,57 @@ protected StreamElement(
     this.entityDescriptor = Objects.requireNonNull(entityDesc);
     this.attributeDescriptor = Objects.requireNonNull(attributeDesc);
     this.uuid = Objects.requireNonNull(uuid);
+    this.sequentialId = -1L;
+    this.key = Objects.requireNonNull(key);
+    this.attribute = Objects.requireNonNull(attribute);
+    this.stamp = stamp;
+    this.value = value;
+    this.deleteWildcard = deleteWildcard;
+  }
+
+  protected StreamElement(
+      EntityDescriptor entityDesc,
+      AttributeDescriptor<?> attributeDesc,
+      long sequentialId,
+      String key,
+      String attribute,
+      long stamp,
+      boolean deleteWildcard,
+      @Nullable byte[] value) {
+
+    this.entityDescriptor = Objects.requireNonNull(entityDesc);
+    this.attributeDescriptor = Objects.requireNonNull(attributeDesc);
+    this.uuid = null;
+    this.sequentialId = sequentialId;
     this.key = Objects.requireNonNull(key);
     this.attribute = Objects.requireNonNull(attribute);
     this.stamp = stamp;
     this.value = value;
     this.deleteWildcard = deleteWildcard;
+
+    Preconditions.checkArgument(
+        sequentialId > 0, "Sequential ID must be greater than zero, got %s", sequentialId);
+  }
+
+  public String getUuid() {
+    if (cachedUuid == null) {
+      if (uuid != null) {
+        cachedUuid = uuid;
+      } else {
+        cachedUuid = key + ":" + attribute + ":" + sequentialId;
+      }
+    }
+    return cachedUuid;
+  }
+
+  public boolean hasSequentialId() {
+    return sequentialId > 0;
   }
 
   @Override
   public String toString() {
     return MoreObjects.toStringHelper(this)
-        .add("uuid", uuid)
+        .add("uuid", getUuid())
```

Comment:
maybe also `sequentialId`?

Reply:
It is "hidden" in the UUID, but we can add it explicitly.